### DISCUSSION TITLE
Meta endpoint to retrieve validation rules

### DISCUSSION THREAD

**Discussion by balexandre** [fastify/fastify#5348] [Ideas]
I'm new to Fastify, and to help the client side (both web and mobile) I've been wondering if we could have a plugin that would automagically create an endpoint for metadata regarding the validation rules of the endpoint

meaning that, what's I'm after is to help the client side to have the data for validation, for example, if the **schema** is something like
```
{
  type: 'object',
  additionalProperties: false,
  properties: {
    name: { type: 'string' },
    description: { type: 'string', pattern: '^.{2,150}$' },
    type: { enum: ['OVERRIDE', 'CUSTOM', 'DEFAULT'] },
  },
  required: ['name'],
}
```

I would love to have a reply to a "automagically" created endpoint that would output something like
```
{
  "name": {
    "meta": {
      "required": true
    }
  },
  "description": {
    "meta": {
      "required": false,
      "regex": "^.{2,150}$"
    }
  },
  "type": {
    "meta": {
      "required": false,
      "values": ["OVERRIDE", "CUSTOM", "DEFAULT"]
    }
  }
}
```

the "automatically" routes could be something like:

if `POST /expenses` exist, then a new route `GET /expenses/meta` would be created, same for `PATCH` and `PUT`s

that way, the client side, could use it for client-side-validation and use enums to fill dropdown for example, not depending on hardcoded values, and they could safely change when the API changes

***

I found this plugin https://github.com/fastify/fastify-response-validation but it's not exactly what I was looking for...

**Comment by jsumners** [comment]
I have done this in the past. The general idea is to:

1. Utilize your own schema container that loads schema files from the file system -- https://fastify.dev/docs/latest/Reference/Server/#schemacontroller
2. Write an endpoint that will serve schemas based on schema id:
```js
'use strict';

module.exports = function getSchema(req, res) {
  const { schemaId } = req.params;
  const schema = this.schemas.getSchema(schemaId, true);
  res.success(schema);
};
```
